src/debug/telemetry/performance_monitor.c

```diff
@@ -397,3 +404,230 @@ int performance_monitor_init(void)
 /* init performance monitor using Zephyr */
 SYS_INIT(performance_monitor_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
 
+#ifdef CONFIG_SOF_TELEMETRY_IO_PERFORMANCE_MEASUREMENTS
+
+SYS_BITARRAY_DEFINE_STATIC(io_performance_data_bit_array, PERFORMANCE_DATA_ENTRIES_COUNT);
+
+#define IO_PERFORMANCE_ALLOC_BYTES 0x1000
+#define IO_PERFORMANCE_MAX_ENTRIES (IO_PERFORMANCE_ALLOC_BYTES / sizeof(struct io_perf_data_item))
+
+static struct io_perf_monitor_ctx perf_monitor_ctx;
+static struct io_perf_data_item io_perf_data_items[IO_PERFORMANCE_MAX_ENTRIES];
+
+int io_perf_monitor_init(void)
+{
+	int ret;
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+
+	k_spinlock_init(&self->lock);
+	k_spinlock_key_t key = k_spin_lock(&self->lock);
+
+	self->io_perf_data = (struct io_perf_data_item *)io_perf_data_items;
+	self->state = IPC4_PERF_MEASUREMENTS_DISABLED;
+
+	ret = perf_bitmap_init(&self->io_performance_data_bitmap, &io_performance_data_bit_array,
+			       IO_PERFORMANCE_MAX_ENTRIES);
+
+	k_spin_unlock(&self->lock, key);
+	return ret;
+}
+
+/* init performance monitor using Zephyr */
+SYS_INIT(io_perf_monitor_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
+
+static struct io_perf_data_item *io_perf_monitor_get_next_slot(struct io_perf_monitor_ctx *self)
+{
+	int idx;
+	int ret;
+	k_spinlock_key_t key = k_spin_lock(&self->lock);
+
+	ret = perf_bitmap_alloc(&self->io_performance_data_bitmap, &idx);
+	if (ret < 0)
+		return NULL;
+	/* ref. FW did not set the bits, but here we do it to not have to use
+	 * isFree() check that the bitarray does not provide yet. Instead we will use isClear
+	 * ,and always set bit on bitmap alloc.
+	 */
+
+	ret = perf_bitmap_setbit(&self->io_performance_data_bitmap, idx);
+	if (ret < 0)
+		return NULL;
+
+	k_spin_unlock(&self->lock, key);
+	return &self->io_perf_data[idx];
+}
+
+int io_perf_monitor_release_slot(struct io_perf_data_item *item)
+{
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+
+	item->is_removed = true;
+
+	int idx = (item - self->io_perf_data) / sizeof(*item);
```

Comment:
Wouldn't this be one place where locking is needed?

Reply:
I added lock for this line alone, we'll see if we get any problems in smoke tests. This function did deadlock when i locked all the code here, but it's probably because bitmap has it's own locks too.